src/components/modebar/index.js

```diff
@@ -12,63 +12,92 @@
- * @Param {object} config.graphInfo  primary plot object containing data and layout
+ * @Param {object} opts
+ * @Param {object} opts.buttons    nested arrays of grouped buttons config objects
+ * @Param {object} opts.container  container div to append modebar
+ * @Param {object} opts.graphInfo  primary plot object containing data and layout
  */
-function ModeBar (config) {
+function ModeBar(opts) {
+    this._snapshotInProgress = false;
+    this.container = opts.container;
+    this.element = document.createElement('div');
 
-    if (!(this instanceof ModeBar)) return new ModeBar();
+    this.update(opts.graphInfo, opts.buttons);
 
-    var _this = this;
+    this.container.appendChild(this.element);
+}
 
-    this._snapshotInProgress = false;
-    this.graphInfo = config.graphInfo;
-    this.element = document.createElement('div');
+var proto = ModeBar.prototype;
+
+/**
+ * Update modebar (buttons and logo)
+ *
+ * @param {object} graphInfo  primary plot object containing data and layout
+ * @param {array of arrays} buttons nested arrays of grouped buttons to initialize
+ *
+ */
+proto.update = function(graphInfo, buttons) {
+    this.graphInfo = graphInfo;
 
-    if(this.graphInfo._context.displayModeBar === 'hover') {
+    var context = this.graphInfo._context;
+
+    if(context.displayModeBar === 'hover') {
         this.element.className = 'modebar modebar--hover';
-    } else {
-        this.element.className = 'modebar';
     }
+    else this.element.className = 'modebar';
+
+    // if buttons or logo have changed, redraw modebar interior
+    var needsNewButtons = !this.hasButtons(buttons),
+        needsNewLogo = (this.hasLogo !== context.displaylogo);
+
+    if(needsNewButtons || needsNewLogo) {
+        this.removeAllButtons();
+
+        this.updateButtons(buttons);
+
+        if(context.displaylogo) {
+            this.element.appendChild(this.getLogo());
+            this.hasLogo = true;
+        }
+    }
+
+    this.updateActiveButton();
+};
 
-    this.buttons = config.buttons;
+proto.updateButtons = function(buttons) {
+    var _this = this;
+
+    this.buttons = buttons;
     this.buttonElements = [];
+    this.buttonsNames = [];
 
-    this.buttons.forEach( function (buttonGroup) {
+    this.buttons.forEach(function(buttonGroup) {
         var group = _this.createGroup();
 
-        buttonGroup.forEach( function (buttonName) {
-            var buttonConfig = _this.config()[buttonName];
-
-            if (!buttonConfig) {
-                throw new Error(buttonName + ' not specfied in modebar configuration');
+        buttonGroup.forEach(function(buttonConfig) {
+            var buttonName = buttonConfig.name;
+            if(!buttonName) {
+                throw new Error('must provide button \'name\' in button config');
+            }
+            if(_this.buttonsNames.indexOf(buttonName) !== -1) {
+                throw new Error('button name \'', + buttonName + '\' is taken');
```

Comment:
Buttons config object must have a unique `name` key, in order to identify them on updates.